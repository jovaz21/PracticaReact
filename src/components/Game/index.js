import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import FrameLayout from '../../layout/FrameLayout';

// Styled Squares
const Square = styled.button`
  background-color: #222;
  color: white;
  width: 20vh;
  height: 20vh;
  font-size: 15vh;
  border: 4px solid #999;
  outline-width: 0;
`;
const squareComponents = [
  Square.extend`
    border-top-width: 0;
    border-left-width: 0;
  `,
  Square.extend`
    border-top-width: 0;
  `,
  Square.extend`
    border-top-width: 0;
    border-right-width: 0;
  `,
  Square.extend`
    border-left-width: 0;
  `,
  Square,
  Square.extend`
    border-right-width: 0;
  `,
  Square.extend`
    border-bottom-width: 0;
    border-left-width: 0;
  `,
  Square.extend`
    border-bottom-width: 0;
  `,
  Square.extend`
    border-bottom-width: 0;
    border-right-width: 0;
  `,
];

// Constants
const EMPTY_SQUARE_VALUE = '\u00A0';

// Game
class Game extends Component {
  // On Click
  onClick(e) {
    if (e.target.childNodes[0].data !== EMPTY_SQUARE_VALUE) {
      return;
    }
    const { onSelect } = this.props;
    onSelect(parseInt(e.target.value, 10));
  }

  // Render Square
  renderSquare(i, SpecificSquare) {
    const { squares } = this.props;
    const SquareInstance = SpecificSquare || Square;
    return (
      <SquareInstance key={i} className={squares[i]} value={i} onClick={e => this.onClick(e)}>
        {squares[i] ? squares[i] : EMPTY_SQUARE_VALUE}
      </SquareInstance>
    );
  }

  // View Render
  render() {
    return (
      <FrameLayout className="squares" orientation="vertical">
        <div keyval="row-1" className="game-row">
          {[0, 1, 2].map(v => this.renderSquare(v, squareComponents[v]))}
        </div>
        <div keyval="row-2" className="game-row">
          {[3, 4, 5].map(v => this.renderSquare(v, squareComponents[v]))}
        </div>
        <div keyval="row-3" className="game-row">
          {[6, 7, 8].map(v => this.renderSquare(v, squareComponents[v]))}
        </div>
      </FrameLayout>
    );
  }
}
Game.propTypes = {
  squares: PropTypes.arrayOf(PropTypes.string).isRequired,
  onSelect: PropTypes.func.isRequired,
};
export default Game;
