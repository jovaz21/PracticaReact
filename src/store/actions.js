export { statusSet as StatusSetAction, statusReset as StatusResetAction } from './status';
export { gameSet as GameSetAction, gameReset as GameResetAction } from './game';
