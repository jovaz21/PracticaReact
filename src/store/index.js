import { createStore } from 'redux';

import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import rootReducer from './reducer';

export { StatusSetAction, StatusResetAction } from './actions';
export { GameSetAction, GameResetAction } from './actions';

/** Make it persisted */
const persistConfig = {
  key: 'root',
  storage,
  stateReconciler: autoMergeLevel2, // see "Merge Process" section for details.
};
const persistedReducer = persistReducer(persistConfig, rootReducer);
/** ****************** */

/* eslint-disable no-undef, no-underscore-dangle */
export const store = createStore(
  persistedReducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
);
/* eslint-enable */

export const persistor = persistStore(store);
