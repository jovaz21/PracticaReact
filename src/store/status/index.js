export { default as statusInitialState } from './initialState';
export { default as statusReducer } from './reducer';
export { statusSet, statusReset } from './actions';
