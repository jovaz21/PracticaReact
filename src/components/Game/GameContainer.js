import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Game from './Game';
import { StatusSetAction, GameSetAction } from '../../store';

// Get Winner
const getWinner = (squares) => {
  let res = null;
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ];
  lines.forEach((line) => {
    const [ix1, ix2, ix3] = line;
    if (squares[ix1] && squares[ix1] === squares[ix2] && squares[ix1] === squares[ix3]) {
      res = squares[ix1];
    }
  });
  return res;
};

// Is Game Finished
const isGameFinished = (winner, squares) => {
  let res = true;

  /* check */
  if (winner != null) {
    return true;
  }

  /* scan */
  squares.forEach((value) => {
    if (value == null) {
      res = false;
    }
  });

  /* done */
  return res;
};

class GameContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      nextPlayer: props.nextPlayer,
      squares: props.squares,
    };
  }

  // On Select
  onSelect(ix) {
    const prevState = this.state;
    const { squares } = prevState;

    /* check */
    if (isGameFinished(getWinner(squares), squares)) {
      return;
    }

    /* set */
    const { doGameSetAction, doStatusSetAction } = this.props;

    /* Set Game */
    const newSquares = squares.slice();
    newSquares[ix] = prevState.nextPlayer;
    doGameSetAction({
      squares: newSquares,
    });

    /* Set Status */
    const nextPlayer = prevState.nextPlayer === 'X' ? 'O' : 'X';
    const curWinner = getWinner(newSquares);
    const isFinished = isGameFinished(curWinner, newSquares);
    doStatusSetAction({
      nextPlayer,
      curWinner,
      isFinished,
    });
  }

  static getDerivedStateFromProps({ nextPlayer, squares }) {
    return {
      nextPlayer,
      squares,
    };
  }

  render() {
    const { squares } = this.state;
    return <Game squares={squares} onSelect={ix => this.onSelect(ix)} />;
  }
}
GameContainer.defaultProps = {
  ...Game.defaultProps,
  onSelect: () => {},
};
GameContainer.propTypes = {
  ...Game.propTypes,
  onSelect: PropTypes.func,
};

const mapStateToProps = state => ({
  nextPlayer: state.status.nextPlayer,
  squares: state.game.squares,
});
const mapDispatchToProps = {
  doStatusSetAction: StatusSetAction,
  doGameSetAction: GameSetAction,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(GameContainer);
