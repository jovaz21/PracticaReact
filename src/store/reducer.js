import { combineReducers } from 'redux';

import { statusReducer } from './status';
import { gameReducer } from './game';

export default combineReducers({
  status: statusReducer,
  game: gameReducer,
});
