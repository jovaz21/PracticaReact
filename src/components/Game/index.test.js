import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import store from '../../store';
import Game from '.';

describe('Component Tests', () => {
  /* set */
  const squares = Array(9).fill(null);
  const component = <Game store={store} nextPlayer="X" squares={squares} />;

  // State Values before/after Click on Button#4
  test('State Values before/after Click on Button#4', () => {
    const wrapper = shallow(component).dive();
    const instance = wrapper.instance();

    /* before */
    expect(wrapper.state('nextPlayer')).toBe('X');

    /* after */
    let gameSetAction = true;
    store.subscribe(() => {
      if (gameSetAction) {
        // Firstly 'GameSet' is expected to be dispatched
        expect(store.getState().game.squares[4]).toBe('X');
        gameSetAction = false;
      } else {
        // ... then 'StatusSet'
        expect(store.getState().status.nextPlayer).toBe('O');
      }
    });
    instance.onSelect(4);
  });

  // Snapshot
  test('Snapshot', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
