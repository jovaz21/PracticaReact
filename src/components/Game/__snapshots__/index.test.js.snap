// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Tests Snapshot 1`] = `
<div
  className="squares sc-bdVaJa iVFlTG"
>
  <div
    className="VerticalLayoutItem game-row"
  >
    <div
      className="game-row"
      keyval="row-1"
    >
      <button
        className="sc-htpNat bLhQoD"
        id="Button0"
        onClick={[Function]}
        value={0}
      >
         
      </button>
      <button
        className="sc-bxivhb kcRBGw"
        id="Button1"
        onClick={[Function]}
        value={1}
      >
         
      </button>
      <button
        className="sc-ifAKCX hCXenb"
        id="Button2"
        onClick={[Function]}
        value={2}
      >
         
      </button>
    </div>
  </div>
  <div
    className="VerticalLayoutItem game-row"
  >
    <div
      className="game-row"
      keyval="row-2"
    >
      <button
        className="sc-EHOje jDqdhO"
        id="Button3"
        onClick={[Function]}
        value={3}
      >
         
      </button>
      <button
        className="X sc-bwzfXH euBMdG"
        id="Button4"
        onClick={[Function]}
        value={4}
      >
        X
      </button>
      <button
        className="sc-bZQynM jaBCQh"
        id="Button5"
        onClick={[Function]}
        value={5}
      >
         
      </button>
    </div>
  </div>
  <div
    className="VerticalLayoutItem game-row"
  >
    <div
      className="game-row"
      keyval="row-3"
    >
      <button
        className="sc-gzVnrw iBncVh"
        id="Button6"
        onClick={[Function]}
        value={6}
      >
         
      </button>
      <button
        className="sc-htoDjs dNcQZT"
        id="Button7"
        onClick={[Function]}
        value={7}
      >
         
      </button>
      <button
        className="sc-dnqmqq kspAgd"
        id="Button8"
        onClick={[Function]}
        value={8}
      >
         
      </button>
    </div>
  </div>
</div>
`;
